### PR TITLE
wip: plugins

### PR DESCRIPTION
**Pull request by maximnaidenov**

### CODE REVIEW

File: src/interface/plugin.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+ */
+Plugin.prototype.onPrepare = function() {
+};
+
+/**
+ * Called before the suite is started
+ * 
+ * @param {String} suite.name - suite name
+ */
+Plugin.prototype.suiteStarted = function(suite) {
```
Comment (tsaleksandrova) [inline src/interface/plugin.js]: for jasmine hooks, what about just setting up jasmine reporters ?

File: src/uiveri5.js [hunk 2]
```diff
@@ -301,258 +301,265 @@ function run(config) {
         });
       };
 
-      // add global matchers
-      beforeEach(function() {
-        jasmine.getEnv().addMatchers(matchers);
-      });
+      // load plugins
+      var plugins = moduleLoader.loadModule('plugins');
+      return Promise.all(plugins.map(function(plugin) {
```
[the remark this answers is not in the document]
Comment (maximnaidenov) [reply src/uiveri5.js]: webdriverjs extended webdriver.promise and it is now a wrapper on native promises ( + execution flow but we don't need it here). 
According https://github.com/angular/protractor/blob/master/lib/config.ts, onPrepare could optionally return a promise.

File: src/interface/plugin.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+Plugin.prototype.setup = function() {
+};
+
+/**
+ * Called after webdriver connection is established but before tests are started.
+ * 
+ * @return {Promise<void>|void} Can return a promise that will be waited to 
+ * be resolved before uiveri5 continues.
+ */
+Plugin.prototype.onPrepare = function() {
```
[the remark this answers is not in the document]
Comment (maximnaidenov) [reply src/interface/plugin.js]: No, it will be called after webdriver connection is established.

File: src/interface/plugin.js (hunk 1 above)
Comment (maximnaidenov) [reply src/interface/plugin.js]: This is exactly how this method will be called. But I would like to keep the plugin api clean and focused so would prefer to encapsulate the jasmine wrapper.

File: src/uiveri5.js [hunk 4]
```diff
@@ -581,9 +596,32 @@ function run(config) {
       return driverActions.mouseMove(bodyElement, {x:-1, y:-1}).perform();
     }
 
+    function _moduleExec(modules,method,args) {
+      return Promise.all(
+        modules.map(function(module) {
+          return module[method].apply(null,args)
```
Comment (tsvetelina-aleksandrova) [inline src/uiveri5.js]: what if the plugin doesn't define all hooks?

Comment (maximnaidenov) [reply src/uiveri5.js]: good catch, will fix this

**Comment by hmanchev** [comment]
Looks good to me. Just correct the commit message and I can approve it.

**Pull request merged**